src/datasets/packaged_modules/imagefolder/imagefolder.py

```diff
@@ -119,44 +170,167 @@ def _split_files_and_archives(self, data_files):
+                                    if metadata_file is None
+                                    and not os.path.relpath(
+                                        downloaded_dir_file, os.path.dirname(downloaded_metadata_file)
+                                    ).startswith("..")
+                                ]
+                                if metadata_file_candidates:
+                                    _, metadata_file, downloaded_metadata_file = min(
+                                        metadata_file_candidates, key=lambda x: count_path_segments(x[0])
+                                    )
+                                    with open(downloaded_metadata_file, "rb") as f:
+                                        pa_metadata_table = paj.read_json(f)
+                                    pa_file_name_array = pa_metadata_table["file_name"]
+                                    pa_file_name_array = pc.replace_substring(
+                                        pa_file_name_array, pattern="\\", replacement="/"
+                                    )
+                                    pa_metadata_table = pa_metadata_table.drop(["file_name"])
+                                    metadata_dir = os.path.dirname(downloaded_metadata_file)
+                                    metadata_dict = {
+                                        file_name: image_metadata
+                                        for file_name, image_metadata in zip(
+                                            pa_file_name_array.to_pylist(), pa_table_to_pylist(pa_metadata_table)
+                                        )
+                                    }
+                            if metadata_dir is not None:
+                                downloaded_dir_file_relpath = os.path.relpath(downloaded_dir_file, metadata_dir)
+                                downloaded_dir_file_relpath = downloaded_dir_file_relpath.replace("\\", "/")
+                                image_metadata = metadata_dict.get(downloaded_dir_file_relpath, image_empty_metadata)
+                            else:
+                                image_metadata = image_empty_metadata
+                            if self.config.drop_labels:
+                                yield file_idx, {
+                                    "image": downloaded_dir_file,
+                                    **image_metadata,
+                                }
+                            else:
+                                yield file_idx, {
+                                    "image": downloaded_dir_file,
+                                    "label": os.path.basename(os.path.dirname(downloaded_dir_file)),
+                                    **image_metadata,
+                                }
                             file_idx += 1
         else:
             file_idx = 0
             for file, downloaded_file_or_dir in files:
                 if file is not None:
```

Comment:
(nit) not a big fan of having a variable named `file` when many other variables also have `file` in their names, could we find something more explicit ? Maybe `original_file` ?